File: frontend/src/lib/recommendation.ts
// =====================================================================
// AI Insurance Policy — recommendation engine
//
// Hybrid approach:
//   1. Rule-based eligibility filtering   (age, family size, coverage,
//                                          budget, pre-existing exclusions)
//   2. Risk classification               (Low / Medium / High)
//   3. Content-based scoring             (compares user profile vector
//                                          with policy feature vector)
//   4. Weighted ranking with the spec'd weights:
//        Budget match        30%
//        Coverage match      25%
//        Claim settlement /
//        company reliability 20%
//        Benefits match      15%
//        Risk suitability    10%
//
// This module is pure (no DB / no I/O) so it can be unit-tested and
// reused on both server and client.
// =====================================================================

export interface UserInput {
  full_name?: string;
  age: number;
  gender: string;
  marital_status?: string;
  occupation?: string;
  family_size: number;
  dependents?: number;
  monthly_income: number;
  monthly_budget: number;
  budget_period?: "monthly" | "yearly";
  city_tier: number;
  smoker: boolean;
  alcohol?: boolean;
  active_lifestyle: boolean;
  lifestyle?: "sedentary" | "average" | "active";
  pre_existing: string[];
  family_history?: string[];
  current_insurance?: string;
  coverage_need: number;
  needed_benefits: string[];
  preferred_policy_type?: string;
}

export interface PolicyRow {
  id: string;
  company_id: string;
  name: string;
  policy_type: string;
  premium_monthly: number;
  coverage_amount: number;
  min_age: number;
  max_age: number;
  max_family_size: number;
  waiting_period_months: number;
  benefits: string[];
  exclusions: string[];
  ideal_age_band_min: number;
  ideal_age_band_max: number;
  claim_settlement_ratio?: number | null;
  network_hospitals?: number | null;
  room_rent_limit?: string;
  co_payment_percentage?: number;
  maternity_cover?: boolean;
  pre_existing_coverage?: boolean;
  pre_existing_waiting_months?: number;
  key_benefits_text?: string;
  policy_score?: number;
  company?: {
    name: string;
    claim_settlement_ratio: number;
    network_hospitals: number;
    customer_rating: number;
  };
}

export interface ScoredPolicy {
  policy: PolicyRow;
  score: number;
  rank: number;
  reason: string;
  factors: Record<string, number>;
}

// Spec weights (must sum to 1.0)
export const WEIGHTS = {
  budget: 0.30,
  coverage: 0.25,
  reliability: 0.20,
  benefits: 0.15,
  risk: 0.10,
} as const;

const clamp = (n: number, min = 0, max = 1) => Math.max(min, Math.min(max, n));

// ---------------------------------------------------------------------
// Monthly-equivalent budget
// ---------------------------------------------------------------------
export function monthlyBudget(user: UserInput): number {
  return user.budget_period === "yearly" ? user.monthly_budget / 12 : user.monthly_budget;
}

// ---------------------------------------------------------------------
// Risk classification (Low / Medium / High)
// Factors: age, pre-existing diseases, smoking, alcohol, lifestyle,
//          family history, occupation, coverage need, current insurance
// ---------------------------------------------------------------------
export function calculateRisk(input: UserInput): {
  level: "Low" | "Medium" | "High";
  score: number;
  factors: Record<string, number>;
} {
  let score = 0;
  const f: Record<string, number> = {};

  // Age
  if (input.age < 30) f.age = 1;
  else if (input.age < 45) f.age = 2;
  else if (input.age < 60) f.age = 4;
  else f.age = 6;
  score += f.age;

  // Pre-existing conditions
  f.pre_existing = input.pre_existing.length * 1.5;
  score += f.pre_existing;

  // Lifestyle: smoking + alcohol + activity
  f.smoking = input.smoker ? 2 : 0;
  f.alcohol = input.alcohol ? 1 : 0;
  if (input.lifestyle === "sedentary") f.activity = 1;
  else if (input.lifestyle === "active") f.activity = -1;
  else f.activity = 0;
  if (input.active_lifestyle) f.activity = Math.min(f.activity, -1);
  score += f.smoking + f.alcohol + f.activity;

  // Family medical history
  f.family_history = (input.family_history?.length ?? 0) * 0.75;
  score += f.family_history;

  // Occupation — flag known high-risk jobs (very basic)
  const hazardous = ["construction", "driver", "miner", "police", "soldier", "firefighter", "factory"];
  const occ = (input.occupation || "").toLowerCase();
  f.occupation = hazardous.some((o) => occ.includes(o)) ? 1 : 0;
  score += f.occupation;

  // Coverage need vs income (high cover ask with low income raises risk)
  const annualIncome = Math.max(input.monthly_income * 12, 1);
  const coverageRatio = input.coverage_need / annualIncome;
  if (coverageRatio > 5) f.coverage_need = 1;
  else if (coverageRatio > 2) f.coverage_need = 0.5;
  else f.coverage_need = 0;
  score += f.coverage_need;

  score = Math.max(0, score);
  const level: "Low" | "Medium" | "High" =
    score <= 3 ? "Low" : score <= 6 ? "Medium" : "High";

  return { level, score: Math.round(score * 10) / 10, factors: f };
}

// ---------------------------------------------------------------------
// Rule-based eligibility filter
// ---------------------------------------------------------------------
export function filterEligible(user: UserInput, policies: PolicyRow[]): PolicyRow[] {
  const budget = monthlyBudget(user);
  return policies.filter((p) => {
    // Age band
    if (user.age < p.min_age || user.age > p.max_age) return false;
    // Family size
    if (user.family_size > p.max_family_size) return false;
    // Coverage need: policy must reach at least 80% of the target.
    if (p.coverage_amount < user.coverage_need * 0.8) return false;
    // Budget: allow up to 15% above stated budget.
    if (p.premium_monthly > budget * 1.15) return false;
    // Pre-existing exclusions
    for (const cond of user.pre_existing) {
      if (p.exclusions.includes(cond)) return false;
    }
    // Preferred policy type (if user picked something specific)
    if (
      user.preferred_policy_type &&
      user.preferred_policy_type !== "any" &&
      p.policy_type !== user.preferred_policy_type
    ) {
      return false;
    }
    return true;
  });
}

// ---------------------------------------------------------------------
// Score & rank — content-based + weighted ranking
// ---------------------------------------------------------------------
export function scorePolicies(user: UserInput, policies: PolicyRow[], risk: "Low" | "Medium" | "High"): ScoredPolicy[] {
  if (policies.length === 0) return [];
  const budget = monthlyBudget(user);

  return policies
    .map((p) => {
      // ---- 1) Budget match ---------------------------------------
      // 1.0 when premium is exactly at budget, falls off symmetrically.
      const budgetDelta = Math.abs(p.premium_monthly - budget) / Math.max(budget, 1);
      const budgetScore = clamp(1 - budgetDelta);

      // ---- 2) Coverage match -------------------------------------
      // Peak at the target; surplus above counts (capped); below penalised.
      const coverageRatio = p.coverage_amount / Math.max(user.coverage_need, 1);
      const coverageScore =
        coverageRatio >= 1 ? clamp(1 - (coverageRatio - 1) * 0.15) : clamp(coverageRatio);

      // ---- 3) Company reliability (claim ratio + network) --------
      const claimRatio = p.claim_settlement_ratio ?? p.company?.claim_settlement_ratio ?? 85;
      const claimScore = clamp((claimRatio - 80) / 20); // 80% → 0, 100% → 1
      const network = p.network_hospitals ?? p.company?.network_hospitals ?? 0;
      const networkScore = clamp(network / 10000); // 10k+ → 1
      const rating = p.company?.customer_rating ?? 4;
      const ratingScore = clamp((rating - 3) / 2); // 3★ → 0, 5★ → 1
      const reliabilityScore = claimScore * 0.6 + networkScore * 0.25 + ratingScore * 0.15;

      // ---- 4) Benefits match -------------------------------------
      const needed = user.needed_benefits.length;
      const overlap = needed === 0 ? 0 : user.needed_benefits.filter((b) => p.benefits.includes(b)).length;
      const benefitsScore = needed === 0 ? 0.7 : overlap / needed;

      // ---- 5) Risk suitability -----------------------------------
      // Match policy to user's risk level.
      //   - High risk users score better on policies with PED coverage,
      //     critical-illness / chronic / senior types, and low waiting.
      //   - Low risk users score better on individual / young-adult plans.
      let riskScore = 0.5;
      const inBand =
        user.age >= p.ideal_age_band_min && user.age <= p.ideal_age_band_max ? 1 : 0;
      if (risk === "High") {
        riskScore =
          (p.pre_existing_coverage ? 0.4 : 0) +
          (p.policy_type === "critical_illness" || p.policy_type === "chronic" || p.policy_type === "senior" ? 0.3 : 0.15) +
          clamp(1 - (p.waiting_period_months ?? 24) / 36) * 0.2 +
          inBand * 0.1;
      } else if (risk === "Medium") {
        riskScore =
          (p.pre_existing_coverage ? 0.25 : 0.1) +
          (p.policy_type === "family_floater" || p.policy_type === "individual" ? 0.3 : 0.2) +
          clamp(1 - (p.waiting_period_months ?? 24) / 36) * 0.2 +
          inBand * 0.25;
      } else {
        riskScore =
          (p.policy_type === "individual" || p.policy_type === "family_floater" ? 0.4 : 0.2) +
          clamp(1 - (p.waiting_period_months ?? 24) / 24) * 0.25 +
          inBand * 0.35;
      }
      riskScore = clamp(riskScore);

      // ---- Weighted total ----------------------------------------
      const factors = {
        budget: budgetScore,
        coverage: coverageScore,
        reliability: reliabilityScore,
        benefits: benefitsScore,
        risk: riskScore,
      };

      const score =
        WEIGHTS.budget * budgetScore +
        WEIGHTS.coverage * coverageScore +
        WEIGHTS.reliability * reliabilityScore +
        WEIGHTS.benefits * benefitsScore +
        WEIGHTS.risk * riskScore;

      return { policy: p, score, rank: 0, reason: "", factors };
    })
    .sort((a, b) => b.score - a.score)
    .map((s, i) => ({ ...s, rank: i + 1, reason: buildReason(user, s) }));
}

function buildReason(user: UserInput, s: ScoredPolicy): string {
  const p = s.policy;
  const parts: string[] = [];
  const inr = (n: number) => `₹${n.toLocaleString("en-IN")}`;
  const lakhs = (n: number) => `₹${(n / 100000).toFixed(1)} lakh`;
  const budget = monthlyBudget(user);
  const claim = p.claim_settlement_ratio ?? p.company?.claim_settlement_ratio ?? 0;
  const net = p.network_hospitals ?? p.company?.network_hospitals ?? 0;

  if (s.factors.budget > 0.75)
    parts.push(`monthly premium of ${inr(p.premium_monthly)} fits well within your ${inr(budget)} budget`);
  else if (s.factors.budget > 0.45)
    parts.push(`premium ${inr(p.premium_monthly)} is close to your ${inr(budget)} budget`);

  if (s.factors.coverage >= 0.95)
    parts.push(`provides ${lakhs(p.coverage_amount)} coverage, matching your ${lakhs(user.coverage_need)} target`);
  else if (s.factors.coverage >= 0.8)
    parts.push(`covers ${lakhs(p.coverage_amount)} — slightly below your ${lakhs(user.coverage_need)} target`);

  if (p.company && claim >= 95)
    parts.push(`${p.company.name} has a strong ${claim}% claim settlement ratio`);
  else if (p.company && claim >= 90)
    parts.push(`${p.company.name} settles ${claim}% of claims`);

  if (net >= 7000) parts.push(`cashless across ${net.toLocaleString("en-IN")}+ hospitals`);

  if (user.needed_benefits.length > 0 && s.factors.benefits >= 0.7) {
    const matched = user.needed_benefits.filter((b) => p.benefits.includes(b));
    if (matched.length > 0) parts.push(`includes the benefits you asked for (${matched.join(", ").replace(/_/g, " ")})`);
  }

  if (p.pre_existing_coverage && user.pre_existing.length > 0)
    parts.push(`covers your pre-existing conditions after ${p.pre_existing_waiting_months ?? 36} months`);

  if (parts.length === 0)
    parts.push(`overall suitability score ${Math.round(s.score * 100)}% across budget, coverage and benefits`);

  return parts.join("; ") + ".";
}

// ---------------------------------------------------------------------
// Top-level entry point
// ---------------------------------------------------------------------
export function runRecommendation(user: UserInput, policies: PolicyRow[]) {
  const risk = calculateRisk(user);
  const eligible = filterEligible(user, policies);
  const scored = scorePolicies(user, eligible, risk.level);
  return {
    scored,
    risk: { level: risk.level, score: risk.score },
    riskFactors: risk.factors,
    eligibleCount: eligible.length,
    totalCount: policies.length,
  };
}

// Convenience re-export for old call sites that used calculateRiskLevel
export const calculateRiskLevel = (u: UserInput) => {
  const r = calculateRisk(u);
  return { level: r.level, score: r.score };
};

File: backend/app/recommendation.py
"""PolicyMate AI — recommendation engine (Python port).

Hybrid pipeline:
  1. Rule-based eligibility filtering
  2. Risk classification (Low / Medium / High)
  3. Content-based scoring
  4. Weighted ranking with spec'd weights:
       Budget match                          30%
       Coverage match                        25%
       Claim settlement / company reliability 20%
       Benefits match                        15%
       Risk suitability                      10%

Pure functions — no DB / no I/O.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal

# Weights (sum = 1.0)
WEIGHTS = {
    "budget": 0.30,
    "coverage": 0.25,
    "reliability": 0.20,
    "benefits": 0.15,
    "risk": 0.10,
}

RiskLevel = Literal["Low", "Medium", "High"]
HAZARDOUS_OCCUPATIONS = (
    "construction", "driver", "miner", "police", "soldier", "firefighter", "factory",
)


def _clamp(n: float, lo: float = 0.0, hi: float = 1.0) -> float:
    return max(lo, min(hi, n))


def monthly_budget(profile: dict) -> float:
    b = float(profile.get("monthly_budget", 0))
    if profile.get("budget_period") == "yearly":
        return b / 12.0
    return b


@dataclass
class RiskResult:
    level: RiskLevel
    score: float
    factors: dict[str, float] = field(default_factory=dict)


def calculate_risk(profile: dict) -> RiskResult:
    factors: dict[str, float] = {}
    score = 0.0

    age = int(profile.get("age", 0))
    if age < 30:
        factors["age"] = 1
    elif age < 45:
        factors["age"] = 2
    elif age < 60:
        factors["age"] = 4
    else:
        factors["age"] = 6
    score += factors["age"]

    factors["pre_existing"] = len(profile.get("pre_existing") or []) * 1.5
    score += factors["pre_existing"]

    factors["smoking"] = 2 if profile.get("smoker") else 0
    factors["alcohol"] = 1 if profile.get("alcohol") else 0

    lifestyle = profile.get("lifestyle", "average")
    if lifestyle == "sedentary":
        factors["activity"] = 1
    elif lifestyle == "active":
        factors["activity"] = -1
    else:
        factors["activity"] = 0
    if profile.get("active_lifestyle"):
        factors["activity"] = min(factors["activity"], -1)
    score += factors["smoking"] + factors["alcohol"] + factors["activity"]

    factors["family_history"] = len(profile.get("family_history") or []) * 0.75
    score += factors["family_history"]

    occ = (profile.get("occupation") or "").lower()
    factors["occupation"] = 1.0 if any(h in occ for h in HAZARDOUS_OCCUPATIONS) else 0.0
    score += factors["occupation"]

    annual_income = max(float(profile.get("monthly_income", 0)) * 12, 1)
    coverage_ratio = float(profile.get("coverage_need", 0)) / annual_income
    if coverage_ratio > 5:
        factors["coverage_need"] = 1.0
    elif coverage_ratio > 2:
        factors["coverage_need"] = 0.5
    else:
        factors["coverage_need"] = 0.0
    score += factors["coverage_need"]

    score = max(0.0, score)
    level: RiskLevel = "Low" if score <= 3 else ("Medium" if score <= 6 else "High")
    return RiskResult(level=level, score=round(score, 1), factors=factors)


def _policy_features(p: Any) -> dict:
    """Reads policy attributes from an ORM Policy object or a dict."""
    if isinstance(p, dict):
        return p

    def g(name: str, default=None):
        return getattr(p, name, default)

    company = g("company")
    return {
        "id": str(g("id")),
        "company_id": str(g("company_id")),
        "name": g("name"),
        "policy_type": g("policy_type"),
        "premium_monthly": float(g("premium_monthly") or 0),
        "coverage_amount": float(g("coverage_amount") or 0),
        "min_age": int(g("min_age") or 0),
        "max_age": int(g("max_age") or 120),
        "max_family_size": int(g("max_family_size") or 6),
        "waiting_period_months": int(g("waiting_period_months") or 0),
        "benefits": list(g("benefits") or []),
        "exclusions": list(g("exclusions") or []),
        "ideal_age_band_min": int(g("ideal_age_band_min") or 18),
        "ideal_age_band_max": int(g("ideal_age_band_max") or 65),
        "claim_settlement_ratio": float(g("claim_settlement_ratio"))
        if g("claim_settlement_ratio") is not None
        else None,
        "network_hospitals": int(g("network_hospitals"))
        if g("network_hospitals") is not None
        else None,
        "room_rent_limit": g("room_rent_limit"),
        "co_payment_percentage": float(g("co_payment_percentage") or 0),
        "maternity_cover": bool(g("maternity_cover")),
        "pre_existing_coverage": bool(g("pre_existing_coverage")),
        "pre_existing_waiting_months": int(g("pre_existing_waiting_months") or 36),
        "key_benefits_text": g("key_benefits_text"),
        "policy_score": float(g("policy_score") or 75),
        "company": {
            "name": getattr(company, "name", None) if company else None,
            "claim_settlement_ratio": float(getattr(company, "claim_settlement_ratio", 85))
            if company
            else 85.0,
            "network_hospitals": int(getattr(company, "network_hospitals", 0)) if company else 0,
            "customer_rating": float(getattr(company, "customer_rating", 4)) if company else 4.0,
        }
        if company
        else None,
    }


def filter_eligible(profile: dict, policies: list[dict]) -> list[dict]:
    budget = monthly_budget(profile)
    out: list[dict] = []
    for p in policies:
        if profile["age"] < p["min_age"] or profile["age"] > p["max_age"]:
            continue
        if profile["family_size"] > p["max_family_size"]:
            continue
        if p["coverage_amount"] < float(profile["coverage_need"]) * 0.8:
            continue
        if p["premium_monthly"] > budget * 1.15:
            continue
        excl = set(p["exclusions"])
        if any(c in excl for c in profile.get("pre_existing") or []):
            continue
        pref = profile.get("preferred_policy_type") or "any"
        if pref != "any" and p["policy_type"] != pref:
            continue
        out.append(p)
    return out


def _build_reason(profile: dict, score: float, factors: dict, p: dict) -> str:
    parts: list[str] = []
    budget = monthly_budget(profile)
    inr = lambda n: f"Rs. {int(n):,}"
    lakhs = lambda n: f"Rs. {n / 100000:.1f} lakh"
    claim = p.get("claim_settlement_ratio") or (
        p.get("company") or {}
    ).get("claim_settlement_ratio") or 0
    net = p.get("network_hospitals") or (p.get("company") or {}).get("network_hospitals") or 0
    company_name = (p.get("company") or {}).get("name") or ""

    if factors["budget"] > 0.75:
        parts.append(
            f"monthly premium of {inr(p['premium_monthly'])} fits well within your {inr(budget)} budget"
        )
    elif factors["budget"] > 0.45:
        parts.append(f"premium {inr(p['premium_monthly'])} is close to your {inr(budget)} budget")

    if factors["coverage"] >= 0.95:
        parts.append(
            f"provides {lakhs(p['coverage_amount'])} coverage, matching your {lakhs(profile['coverage_need'])} target"
        )
    elif factors["coverage"] >= 0.8:
        parts.append(
            f"covers {lakhs(p['coverage_amount'])} — slightly below your {lakhs(profile['coverage_need'])} target"
        )

    if company_name and claim >= 95:
        parts.append(f"{company_name} has a strong {claim}% claim settlement ratio")
    elif company_name and claim >= 90:
        parts.append(f"{company_name} settles {claim}% of claims")

    if net >= 7000:
        parts.append(f"cashless across {net:,}+ hospitals")

    needed = profile.get("needed_benefits") or []
    if needed and factors["benefits"] >= 0.7:
        matched = [b for b in needed if b in p["benefits"]]
        if matched:
            parts.append(
                "includes the benefits you asked for ("
                + ", ".join(b.replace("_", " ") for b in matched)
                + ")"
            )

    if p["pre_existing_coverage"] and profile.get("pre_existing"):
        parts.append(
            f"covers your pre-existing conditions after {p.get('pre_existing_waiting_months', 36)} months"
        )

    if not parts:
        parts.append(
            f"overall suitability score {round(score * 100)}% across budget, coverage and benefits"
        )
    return "; ".join(parts) + "."


def score_policies(profile: dict, policies: list[dict], risk_level: RiskLevel) -> list[dict]:
    """Returns a sorted list of dicts: {policy, score, factors, rank, reason}."""
    budget = monthly_budget(profile)
    scored: list[dict] = []

    for p in policies:
        # Budget
        delta = abs(p["premium_monthly"] - budget) / max(budget, 1)
        budget_score = _clamp(1 - delta)

        # Coverage
        ratio = p["coverage_amount"] / max(profile["coverage_need"], 1)
        coverage_score = _clamp(1 - (ratio - 1) * 0.15) if ratio >= 1 else _clamp(ratio)

        # Reliability
        claim = p.get("claim_settlement_ratio") or (p.get("company") or {}).get(
            "claim_settlement_ratio", 85
        )
        claim_score = _clamp((float(claim) - 80) / 20)
        net = p.get("network_hospitals") or (p.get("company") or {}).get("network_hospitals", 0)
        net_score = _clamp(float(net) / 10000)
        rating = (p.get("company") or {}).get("customer_rating", 4)
        rating_score = _clamp((float(rating) - 3) / 2)
        reliability_score = claim_score * 0.6 + net_score * 0.25 + rating_score * 0.15

        # Benefits
        needed = profile.get("needed_benefits") or []
        if not needed:
            benefits_score = 0.7
        else:
            overlap = sum(1 for b in needed if b in p["benefits"])
            benefits_score = overlap / len(needed)

        # Risk suitability
        in_band = (
            1.0 if profile["age"] >= p["ideal_age_band_min"] and profile["age"] <= p["ideal_age_band_max"]
            else 0.0
        )
        waiting_score = _clamp(1 - (p["waiting_period_months"] or 0) / 36)
        if risk_level == "High":
            risk_score = (
                (0.4 if p["pre_existing_coverage"] else 0)
                + (0.30 if p["policy_type"] in ("critical_illness", "chronic", "senior") else 0.15)
                + waiting_score * 0.2
                + in_band * 0.1
            )
        elif risk_level == "Medium":
            risk_score = (
                (0.25 if p["pre_existing_coverage"] else 0.10)
                + (0.30 if p["policy_type"] in ("family_floater", "individual") else 0.20)
                + waiting_score * 0.2
                + in_band * 0.25
            )
        else:  # Low
            risk_score = (
                (0.40 if p["policy_type"] in ("individual", "family_floater") else 0.20)
                + _clamp(1 - (p["waiting_period_months"] or 0) / 24) * 0.25
                + in_band * 0.35
            )
        risk_score = _clamp(risk_score)

        factors = {
            "budget": round(budget_score, 4),
            "coverage": round(coverage_score, 4),
            "reliability": round(reliability_score, 4),
            "benefits": round(benefits_score, 4),
            "risk": round(risk_score, 4),
        }
        total = (
            WEIGHTS["budget"] * budget_score
            + WEIGHTS["coverage"] * coverage_score
            + WEIGHTS["reliability"] * reliability_score
            + WEIGHTS["benefits"] * benefits_score
            + WEIGHTS["risk"] * risk_score
        )

        scored.append({"policy": p, "score": round(total, 6), "factors": factors})

    scored.sort(key=lambda s: s["score"], reverse=True)
    for i, s in enumerate(scored):
        s["rank"] = i + 1
        s["reason"] = _build_reason(profile, s["score"], s["factors"], s["policy"])
    return scored


def run_recommendation(profile: dict, policies_orm: list[Any]) -> dict:
    """Top-level entry — accepts a profile dict and a list of ORM Policy rows."""
    policies = [_policy_features(p) for p in policies_orm]
    risk = calculate_risk(profile)
    eligible = filter_eligible(profile, policies)
    scored = score_policies(profile, eligible, risk.level)
    return {
        "scored": scored,
        "risk": {"level": risk.level, "score": risk.score},
        "risk_factors": risk.factors,
        "eligible_count": len(eligible),
        "total_count": len(policies),
    }
